### PR TITLE
Fix Issue 16657 and 11161 - alias this interacts with generated opCmp and opEquals

### PR DESCRIPTION
**Pull request by RazvanN7**
A few exaplanations: it seems that there is a long history of wrong decisions regarding how alias this should be employed. For more information: 

[1] - makes it so that the alias this is first checked in comparisons
[2] - further deepens the mistake in [1]
[3] - segmentation fault due to [1], fixed by issuing an error, although the code could fallback to bitwise comparison

This patch fixes both issues 16657 and 11161 and, in my humble opinion, puts the first brick on the right path. In order to do so, I had to modify the tests for the above mentioned issues.

[1] https://issues.dlang.org/show_bug.cgi?id=10037
[2] https://issues.dlang.org/show_bug.cgi?id=14806
[3] https://issues.dlang.org/show_bug.cgi?id=15292

### CODE REVIEW

**Comment by RazvanN7** [comment]
Currently blocked by https://github.com/dlang/druntime/pull/2476

**Comment by jacob-carlborg** [comment]
Is this a breaking change since `Range` [1] had to be patched?

[1] https://github.com/dlang/druntime/pull/2476

**Comment by rainers** [comment]
Is there a comprehensive list of expected lookup precedence rules somewhere? I'm thinking of current and parent scopes, base classes, alias this, opDispatch, legacy operator overloads, imports, template mixin, UFCS functions and probably more. If it's not specified in detail in the spec it will probably move to and fro depending on the last raised use case.

**Comment by andralex** [comment]
@jacob-carlborg it is changing behavior, however that was arguably accept-invalid...

**Comment by andralex** [comment]
I don't see a reasonable deprecation path here - it would be onerous to require folks who define alias this to also define opEquals... I'm lost here. @WalterBright ?

**Review by WalterBright — APPROVED** [review]

**Comment by UplinkCoder** [comment]
Where's the spec change?

**Comment by 12345swordy** [comment]
What should be the deprecation path?  This undoubtedly going to break code.

**Comment by RazvanN7** [comment]
I added a changelog entry and I made a PR to modify the spec.

I don't see how we could use a deprecation cycle here. This is changing behavior from code that choose an `opEquals` to code that chooses a different `opEquals`; both methods always exist. It seems that the projects listed in buildkite don't suffer from this bug so I suggest we just swallow this pill and merge this.

**Comment by jacob-carlborg** [comment]
1. Does this mean that `opEquals` will never be used through an `alias this` for structs?
2. What was the behavior before this PR when `opEquals` was explicitly defined?

**Comment by RazvanN7** [comment]
>Does this mean that opEquals will never be used through an alias this for structs?

It will be if the generated opEquals does not match. Consider:

```d
struct A
{
    int a;
}

struct B
{
    A a;
    int b;
    alias a this;
}

void main()
{
    A a;
    B b;
    assert(b == a); // gets lowered to b.s.opEquals(a) ;    
}
```

This example is also present in the spec change. You can check it for more information.

>What was the behavior before this PR when opEquals was explicitly defined

Before this PR, the alias this `opEquals` would have been preferred instead of the `struct`
generated one. However, if a user-defined `opEquals` is present, it will be used instead of the alias this one.

**Comment by jacob-carlborg** [comment]
Regarding deprecation path. You could print a deprecation message if a struct has an `alias this` and the `alias this` target explicitly defines `opEquals`.

**Comment by RazvanN7** [comment]
But that will be a deprecation that will never turn into an error. It would probably be more suited as a warning, but you know Walter's stance on warnings.

**Comment by jacob-carlborg** [comment]
Yes but I don’t see how a silent change is better.

File: changelog/alias_this_opEquals.dd [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+equality comparison (`==`), the aliased this one is wrongfully preferred.
+
+This patch corrects this issue by always choosing the defined `opEquals` (even if it is
+generated by the compiler) instead of the alias this one. If the behavior prior to this
+patch is desired, an explicit `opEquals` needs to be provided.
+
+References:
+[1] https://issues.dlang.org/show_bug.cgi?id=16657
+[2] https://github.com/dlang/dmd/pull/9289
+[3] https://github.com/dlang/dlang.org/pull/2593
```
Comment (wilzbach) [inline changelog/alias_this_opEquals.dd]: This doesn't play well with Ddoc - see DAutoTest.

**Comment by wilzbach** [comment]
>  It seems that the projects listed in buildkite don't suffer from this bug so I suggest we just swallow this pill and merge this.

:+1:

Let's move forward with this. We need to fix the status quo.

**Review by thewilsonator — APPROVED** [review]

**Comment by JohanEngelen** [comment]
The way these kind of changes are handled hurt professional development with D _a lot_. Not having a warning/deprecation message about this change is very disturbing. I basically cannot trust the language, ever, and this PR just adds to the growing mountain of evidence...

The change may have been for the better, great, let's make the change!
**But please, make the compiler complain about this (several good solutions here) such that the user notices significant changes in language behavior.**

**Comment by UplinkCoder** [comment]
@JohanEngelen I feel you.
I've tried to formalized a part of D but I can't do it.
It's a moving target.

**Comment by CyberShadow** [comment]
Link to the forum discussion regarding the above:
https://forum.dlang.org/post/dsklfejzyfmytckeuhby@forum.dlang.org

**Pull request merged**